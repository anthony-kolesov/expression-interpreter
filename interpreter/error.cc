/* Code to report errors to the user.
 * Copyright (C) 2017 Anton Kolesov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "error.h"

int user_error(const YYLTYPE *loc, const std::string &msg) {
    std::cerr << "ERROR:" << loc->first_line << "," << loc->first_column;
    if (loc->first_line != loc->last_line
            || loc->first_column != loc->last_column) {
        std::cerr << "-" << loc->last_line << "," << loc->last_column;
    }
    std::cerr << ":" << msg << std::endl;
}

int user_error(int line, const std::string &msg) {
    std::cerr << "ERROR:" << line << ":" << msg << std::endl;
}

// vim: tabstop=4 softtabstop=4 shiftwidth=4 expandtab
