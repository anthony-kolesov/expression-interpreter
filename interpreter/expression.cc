/* Interpreter expressions.
 * Copyright (C) 2017 Anton Kolesov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "expression.h"

std::vector<ValuePtr> MapExpression::getResult(
    ValuePtr input, const std::string &paramName,
    std::shared_ptr<const Expression> func) {

    std::vector<ValuePtr> seq;
    seq.reserve(input->getSize());
    Context funcCtx;
    for (; !input->isNone(); input = input->next()) {
        funcCtx.setVariable(paramName, input->asScalar());
        auto newValue = func->evaluate(&funcCtx);
        seq.push_back(newValue);
    }
    return seq;
}

ValuePtr MapExpression::evaluate(Context *ctx) const {
    auto inputVal = this->input_->evaluate(ctx);
    auto inputSize = inputVal->getSize();
    auto slicesCount = std::min<int>(inputSize,
                                     std::thread::hardware_concurrency());
    auto begin = 0, end = 0;
    auto sliceSize = inputSize / slicesCount;
    std::vector< std::future< std::vector<ValuePtr> > > intermediate;

    for (auto i = 0; i < slicesCount; i++) {
        begin = end;
        /* Make sure we don't last elements in last slice, in case inputSize
         * doesn't divide evenly.  */
        if (i == slicesCount - 1) {
            end = inputSize;
        } else {
            end += sliceSize;
        }

        auto future = std::async(std::launch::async, getResult,
                                 inputVal->getSlice(begin, end),
                                 this->paramName_, this->func_);
        intermediate.push_back(std::move(future));
    }

    auto vector = new std::vector<ValuePtr>();
    vector->reserve(inputSize);
    for (auto &&future : intermediate) {
        auto slice = future.get();
        vector->insert(vector->end(), slice.begin(), slice.end());
    }
    return std::make_shared<const VectorValue>(vector);
}

ValuePtr ReduceExpression::getResult(ValuePtr input, ValuePtr dflt,
                                     const std::string &param1,
                                     const std::string &param2,
                                     std::shared_ptr<const Expression> func) {
    Context funcCtx;
    auto result = dflt;
    for (; !input->isNone(); input = input->next()) {
        funcCtx.setVariable(param1, result);
        funcCtx.setVariable(param2, input->asScalar());
        result = func->evaluate(&funcCtx);
    }
    return result;
}

ValuePtr ReduceExpression::evaluate(Context *ctx) const {
    auto inputVal = this->input_->evaluate(ctx);
    auto dflt = default_->evaluate(ctx);
    auto inputSize = inputVal->getSize();
    auto slicesCount = std::min<int>(inputSize,
                                     std::thread::hardware_concurrency());

    std::vector< std::future< ValuePtr > > intermediate;
    auto begin = 0, end = 0;
    auto sliceSize = inputSize / slicesCount;

    for (auto i = 0; i < slicesCount; i++) {
        begin = end;
        /* Make sure we don't last elements in last slice, in case inputSize
         * doesn't divide evenly.  */
        if (i == slicesCount - 1) {
            end = inputSize;
        } else {
            end += sliceSize;
        }

        auto future = std::async(std::launch::async, getResult,
                                 inputVal->getSlice(begin, end),
                                 dflt, this->param1Name_, this->param2Name_,
                                 this->func_);
        intermediate.push_back(std::move(future));
    }

    auto intermediateValues = new std::vector<ValuePtr>();
    intermediateValues->reserve(slicesCount);
    for (auto &&future : intermediate) {
        intermediateValues->push_back(future.get());
    }

    auto finalValue = std::make_shared<const VectorValue>(intermediateValues);
    return getResult(finalValue, dflt, this->param1Name_, this->param2Name_,
                     this->func_);
}

// vim: tabstop=4 softtabstop=4 shiftwidth=4 expandtab
