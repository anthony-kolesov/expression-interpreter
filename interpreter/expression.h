/* Definition of the structure used to build the syntax tree.
 * Copyright (C) 2017 Anton Kolesov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef EXPRESSION_H_
#define EXPRESSION_H_

#include "context.h"
#include "value.h"

/**
 * @brief Interpeter expression.
 */
class Expression {
 public:
    Expression() { }

    virtual ~Expression() { }

    /**
     * @brief Evaluate value of this expression.
     */
    virtual Value evaluate(Context *ctx) const = 0;
};

class AddExpression : public Expression {
 private:
    Expression *left_;
    Expression *right_;

 public:
    AddExpression(Expression* left, Expression* right) {
        this->left_ = left;
        this->right_ = right;
    }

    virtual ~AddExpression() {
        delete this->left_;
        delete this->right_;
    }

    virtual Value evaluate(Context *ctx) const {
        return this->left_->evaluate(ctx) + this->right_->evaluate(ctx);
    }
};

class DivExpression : public Expression {
 private:
    Expression *left_;
    Expression *right_;

 public:
    DivExpression(Expression* left, Expression* right) {
        this->left_ = left;
        this->right_ = right;
    }

    virtual ~DivExpression() {
        delete this->left_;
        delete this->right_;
    }

    virtual Value evaluate(Context *ctx) const {
        return this->left_->evaluate(ctx) / this->right_->evaluate(ctx);
    }
};

class IdentifierExpression : public Expression {
 private:
    std::string identifier_;

 public:
    explicit IdentifierExpression(const std::string &identifier) {
        this->identifier_ = identifier;
    }

    virtual Value evaluate(Context *ctx) const {
        return ctx->getVariable(this->identifier_);
    }
};

class MulExpression : public Expression {
 private:
    Expression *left_;
    Expression *right_;

 public:
    MulExpression(Expression* left, Expression* right) {
        this->left_ = left;
        this->right_ = right;
    }

    virtual ~MulExpression() {
        delete this->left_;
        delete this->right_;
    }

    virtual Value evaluate(Context *ctx) const {
        return this->left_->evaluate(ctx) * this->right_->evaluate(ctx);
    }
};

class PowExpression : public Expression {
 private:
    Expression *left_;
    Expression *right_;

 public:
    PowExpression(Expression* left, Expression* right) {
        this->left_ = left;
        this->right_ = right;
    }

    virtual ~PowExpression() {
        delete this->left_;
        delete this->right_;
    }

    virtual Value evaluate(Context *ctx) const {
        return this->left_->evaluate(ctx).pow(this->right_->evaluate(ctx));
    }
};

class RangeExpression : public Expression {
 private:
    Expression *begin_;
    Expression *end_;

 public:
    RangeExpression(Expression* begin, Expression* end) : begin_(begin),
            end_(end) { }

    virtual ~RangeExpression() {
        delete this->begin_;
        delete this->end_;
    }

    virtual Value evaluate(Context *ctx) const {
        int begin  = this->begin_->evaluate(ctx).asInteger();
        int end = this->end_->evaluate(ctx).asInteger();
        return Value(begin, end);
    }
};

class SubExpression : public Expression {
 private:
    Expression *left_;
    Expression *right_;

 public:
    SubExpression(Expression* left, Expression* right) {
        this->left_ = left;
        this->right_ = right;
    }

    virtual ~SubExpression() {
        delete this->left_;
        delete this->right_;
    }

    virtual Value evaluate(Context *ctx) const {
        return this->left_->evaluate(ctx) - this->right_->evaluate(ctx);
    }
};

class ValueExpression : public Expression {
 private:
    Value value_;

 public:
    explicit ValueExpression(const Value &v) {
        this->value_ = v;
    }

    virtual Value evaluate(Context *ctx) const {
        return this->value_;
    }
};

#endif  // EXPRESSION_H_

// vim: tabstop=4 softtabstop=4 shiftwidth=4 expandtab
