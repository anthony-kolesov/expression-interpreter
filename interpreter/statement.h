/* Implementation of possible program statements.
 * Copyright (C) 2017 Anton Kolesov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef STATEMENT_H_
#define STATEMENT_H_

#include <iostream>

#include "context.h"
#include "expression.h"

/**
 * @brief Abstract class to represent program statements
 */
class Statement {
 public:
    /**
     * @brief Execute statement.
     */
    virtual void execute(Context *ctx) = 0;
};

/**
 * @brief Statements that print expression result to output.
 */
class OutStatement : public Statement {
 private:
    const Expression *expr_;

 public:
    explicit OutStatement(const Expression *expr) : expr_(expr) { }

    ~OutStatement() {
        delete expr_;
    }

    virtual void execute(Context *ctx) {
        std::cout << this->expr_->evaluate(ctx);
    }
};

class PrintStatement : public Statement {
 private:
    const std::string str_;

 public:
    explicit PrintStatement(const std::string &s) : str_(s) { }

    virtual void execute(Context *ctx) {
        std::cout << this->str_;
    }
};

class VarStatement : public Statement {
 private:
    const std::string name_;
    const Expression *expr_;

 public:

    VarStatement(const std::string &name, const Expression *expr) :
            name_(name), expr_(expr) { }

    ~VarStatement() {
        delete this->expr_;
    }

    virtual void execute(Context *ctx) {
        ctx->setVariable(this->name_, this->expr_->evaluate(ctx));
    }
};

#endif  // STATEMENT_H_

// vim: tabstop=4 softtabstop=4 shiftwidth=4 expandtab
