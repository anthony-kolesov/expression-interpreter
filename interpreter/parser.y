%{

/*
 * parser.y file
 * To generate the parser run: "bison parser.y"
 */

#include "expression.h"
#include "error.h"
#include "statement.h"
#include "parser.h"
#include "lexer.h"

int yyerror(const YYLTYPE *loc, Statement **statement, yyscan_t scanner,
        const char *msg) {
    user_error(loc, std::string(msg));
}

%}

%code requires {

#ifndef YY_TYPEDEF_YY_SCANNER_T
#define YY_TYPEDEF_YY_SCANNER_T
typedef void* yyscan_t;
#endif

}

%output  "parser.c"
%defines "parser.h"

%define api.pure
%lex-param   { yyscan_t scanner }
%parse-param { Statement **statement }
%parse-param { yyscan_t scanner }

%locations

%union {
    int value;
    Expression *expression;
    Statement *statement;
}

%left '+' TOKEN_PLUS
%left '*' TOKEN_MULTIPLY

%token TOKEN_LPAREN
%token TOKEN_RPAREN
%token TOKEN_PLUS
%token TOKEN_MULTIPLY
%token <value> TOKEN_NUMBER
%token TOKEN_OUT

%type <expression> expr
%type <statement> stmt

%%

input
    : stmt { *statement = $1; }
    |
    ;

stmt
    : TOKEN_OUT expr[E] { $$ = new OutStatement($E); }
    ;

expr
    : expr[L] TOKEN_PLUS expr[R] { $$ = new Expression( ePLUS, $L, $R ); }
    | expr[L] TOKEN_MULTIPLY expr[R] { $$ = new Expression( eMULTIPLY, $L, $R ); }
    | TOKEN_LPAREN expr[E] TOKEN_RPAREN { $$ = $E; }
    | TOKEN_NUMBER { $$ = new Expression($1); }
    ;

%%

// vim: tabstop=4 softtabstop=4 shiftwidth=4 expandtab
