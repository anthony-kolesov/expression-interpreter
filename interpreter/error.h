/* Code to report errors to the user.
 * Copyright (C) 2017 Anton Kolesov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef ERROR_H
#define ERROR_H

#include <iostream>
#include <string>

#include "expression.h"
#include "statement.h"
#include "parser.h"

/**
 * @brief Report an error to the user in format "ERROR:<line>,<column>:msg".
 *
 * @param loc Location of error in code.  last_{line,column} are printed only
 * when different from first_{line,column}.
 * @param msg Error text to show to the user.
 */
int user_error(const YYLTYPE *loc, const std::string &msg);

#endif // ERROR_H
