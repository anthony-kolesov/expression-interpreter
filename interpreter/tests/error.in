out ABC
out out
4 + 2
wrong out 78
out 5@2
out 5.4
# This will not be printed.
out 45
print "\n"
# This will not create an error message, because it's a runtime error.
out ABC
