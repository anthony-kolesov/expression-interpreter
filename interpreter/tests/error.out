ERROR:0,4:Unknown symbol `A'
ERROR:0,4:syntax error, unexpected $undefined, expecting TOKEN_LPAREN or TOKEN_NUMBER
ERROR:1,4-1,6:syntax error, unexpected TOKEN_OUT, expecting TOKEN_LPAREN or TOKEN_NUMBER
ERROR:2,0:syntax error, unexpected TOKEN_NUMBER, expecting $end
ERROR:3,0:Unknown symbol `w'
ERROR:3,0:syntax error, unexpected $undefined, expecting $end
ERROR:4,5:Unknown symbol `@'
ERROR:4,5:syntax error, unexpected $undefined, expecting $end
ERROR:5,5:Unknown symbol `.'
ERROR:5,5:syntax error, unexpected $undefined, expecting $end
45
ERROR:7,10:Unknown symbol `A'
ERROR:7,10:syntax error, unexpected $undefined, expecting TOKEN_LPAREN or TOKEN_NUMBER
